### DISCUSSION TITLE
Is TripleDES supported with FIPS provider, error - error:1C8000A8:Provider routines::invalid padding mode

### DISCUSSION THREAD

**Discussion by ChauhanPriyanka** [openssl/openssl#26187] [Q&A]
Hi team, 

Want to understand if I can use Triple DES (3DES) with FIPS provider.
I'm using PKCS7_encrypt with 3DES, getting following error - _error:10880021:PKCS7 routines::PKCS7 lib_

Regards,
Priyanka

**Comment by paulidale** [comment]
The 3.0 FIPS provider includes one 3DES mode (ECB).  This is unlikely to be enough for what you are trying to do.

**Reply by ChauhanPriyanka** [reply]
On updating the algorithm to 3des ECB, I'm getting the following error - error:10800090:PKCS7 routines::cipher has no object identifier

**Reply by t8m** [reply]
ECB mode should not and usually cannot be used in place of other cipher modes.

**Reply by ChauhanPriyanka** [reply]
ok, so I changed to aes_128_CBC, I'm getting error:1C8000A8:Provider routines::invalid padding mode.

**Reply by t8m** [reply]
However the 3.0 FIPS provider also supports the CBC mode.

The newer versions still support it but with fips=no property as it is not approved anymore.

**Reply by ChauhanPriyanka** [reply]
In my scenario, with fips provider what is the best suitable alternate for 3des cbc

**Reply by t8m** [reply]
aes-128-cbc or aes-256-cbc would be the best alternative for 3des-cbc. Not sure why are you getting invalid padding mode error.

**Reply by t8m** [reply]
BTW the invalid padding mode comes from RSA - apparently your code is doing something wrong with an RSA key.

**Reply by ChauhanPriyanka** [reply]
Code was working fine with 3DES, I simply changed 3des to aes-128-cbc & started getting this error - error:1C8000A8:Provider routines::invalid padding mode. RSA Key was generated the same way it was before. I'm reviewing further.

**Reply by ChauhanPriyanka** [reply]
From _fips_module.pod_, I see following documentation, can this error `error:1C8000A8:Provider routines::invalid padding mode ` be related as it is pointing to provider routine.

```
=item RSA Encryption
"pkcs1" padding is no longer approved. 
```

**Reply by ChauhanPriyanka** [reply]
Seems this is related - https://github.com/openssl/openssl/pull/25070.

**Reply by t8m** [reply]
Ah, yes, that's true. RSA encryption with PKCS1 v1.5 padding is not FIPS-approved anymore.

**Reply by ChauhanPriyanka** [reply]
please guide me how to disable padding, not sure if this is a secure way, or if there is any alternate way. The change is breaking PKCS7_encrypt api in fips mode, how to address this.

**Reply by t8m** [reply]
Unfortunately PKCS7 API does not allow changing the encryption padding mode.

It is possible to change it with the CMS API which is a replacement for PKCS7 API. You can use CMS_RecipientInfo_get0_pkey_ctx() to obtain the EVP_PKEY_CTX context and set the padding there with EVP_PKEY_CTX_set_rsa_padding().

**Reply by ChauhanPriyanka** [reply]
I'm using PKCS7 encrypted data in CSR of SCEP request.  So are you saying I can use CMS encrypt here. I need to modify both encypt and decrypt code. With my limited knowledge I'm not able to assess the impact of the change. Can I say cms api can be used interchangeably with PKCS7 api for encrypt/decrypt operations.

**Reply by ChauhanPriyanka** [reply]
ok, CMS is derived from PKCS7 https://www.rfc-editor.org/rfc/rfc5652

**Reply by ChauhanPriyanka** [reply]
I'm not able to get the key context from reciepient_info, there are no openssl errors, . Could you please assist
Do I need to set the public key in recipient_info from X509 cert and then get the context to update padding.
```
CMS_RecipientInfo* recipient_info = sk_CMS_RecipientInfo_value(recipients, 0);
EVP_PKEY_CTX *pKeyCtx = CMS_RecipientInfo_get0_pkey_ctx(recipient_info);
if (pKeyCtx == NULL)
{
   handleErrors();
}
```

**Reply by t8m** [reply]
Did you try to call CMS_encrypt_ex() first and then CMS_add1_recipient() to get the recipient_info?

**Reply by ChauhanPriyanka** [reply]
yes, added code below. Also at the end do I need to call cms_final ?
```
	CMS_ContentInfo* cms = CMS_encrypt_ex(NULL, encbio, EVP_aes_128_cbc(), CMS_BINARY, NULL, NULL);
	if (cms == NULL)
	{
		handleErrors();
		goto cleanup;
	}
	for (int i = 0; i < sk_X509_num(recipients); i++)
	{
		CMS_RecipientInfo* ri;
		EVP_PKEY_CTX* pctx;
		X509* cert = sk_X509_value(recipients, i);
		int res;

		ri = CMS_add1_recipient_cert(cms, cert, 0); // do i need add any specific flag here to store the key context ??
		if (ri == NULL)
		{
			handleErrors();
			goto cleanup;
		}

		pctx = CMS_RecipientInfo_get0_pkey_ctx(ri);
		if (pctx == NULL)
		{
			handleErrors();   //failing here
			goto cleanup;
		}
		if (EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_OAEP_PADDING) <= 0) {
			fprintf(stderr, "Error setting RSA padding mode\n");
			return;
		}
	}
```

**Reply by ChauhanPriyanka** [reply]
I'm able to get the key context, by updating the flag, however beacuse of my limited knowledge I'm not sure how to use the key now. Could you please review do I need to call CMS_Final, how the key will be used for encryption.

> ri = CMS_add1_recipient_cert(cms, cert, CMS_KEY_PARAM);

**Reply by t8m** [reply]
Yes. CMS_Final will perform the encryption.

**Reply by ChauhanPriyanka** [reply]
I'm getting error in CMS_Final error:1C80006B:Provider routines::wrong final block length
```
       inpMessageLen= BIO_get_mem_data(databio, &inpMessage);
	if((encbio	= BIO_new_mem_buf(inpMessage, inpMessageLen)) ==  NULL)
	{
		ReportAPIError("[PKCS7_Wrap] Failed to crete the encryption bio", szErr);
		goto cleanup;
	}
	if (CMS_final(cms, encbio, NULL, CMS_STREAM) == 0)
	{
		handleErrors();
		goto cleanup;
	}
```

**Reply by ChauhanPriyanka** [reply]
The error is same - error:1C80006B:Provider routines::wrong final block length. If I do not set the RSA padding or disable it.

**Reply by t8m** [reply]
I do not think you should use the CMS_STREAM flag.

**Reply by ChauhanPriyanka** [reply]
ok, even after removing CMS_STREAM flag the error is same.

**Reply by ChauhanPriyanka** [reply]
Sharing the code that is working now, but I don't know how it is working. By working I mean, there are no openssl error.

- Replace CMS_encrypt_ex with CMS_encrypt and add flag `CMS_STREAM `in CMS_encrypt- **Working**
- Remove the flag `CMS_STREAM ` from  CMS_encrypt- **Not Working**
- Add the CMS_Stream flag in CMS_final. **Not Working**

Error is same - error:1C80006B:Provider routines::wrong final block length.

Could you please help me to understand why it is working in this specific combination. 

```
CMS_ContentInfo* cms = CMS_encrypt(NULL, encbio, scepReq->scep->conf->enc_alg, CMS_STREAM);
if (cms == NULL)
{
	handleErrors();
	goto cleanup;
}
for (int i = 0; i < sk_X509_num(recipients); i++)
{
	CMS_RecipientInfo* ri;
	EVP_PKEY_CTX* pctx;
	X509* cert = sk_X509_value(recipients, i);
	int res;
	ri = CMS_add1_recipient_cert(cms, cert, CMS_KEY_PARAM);
	if (ri == NULL)
	{
		handleErrors();
		goto cleanup;
	}
	pctx = CMS_RecipientInfo_get0_pkey_ctx(ri);
	if (pctx == NULL)
	{
		handleErrors();
		goto cleanup;
	}
	if (EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_OAEP_PADDING) == 0) {
		handleErrors();
		fprintf(stderr, "Error setting RSA padding mode\n");
		return;
	}
	int mode = 0;
	EVP_PKEY_CTX_get_rsa_padding(pctx, &mode);
	if (mode == 0)
	{
		handleErrors();
		goto cleanup;
	}
}
if (CMS_final(cms, encbio, NULL, 0) == 0)
{
	handleErrors();
	goto cleanup;
}
```

**Comment by ChauhanPriyanka** [comment]
Able to update the RSA padding using CMS API, hence closing the discussion
